### PR TITLE
feat(ui): adds HeaderContainer component

### PR DESCRIPTION
**Pull request by MartinS-git**
# Summary

To enable <TopNavigation> to behave like <PageHeader>, a container component was created: <HeaderContainer>, which is placed in the <AppShell>. The result is that <TopNavigation> is now also sticky and always remains at the top when scrolling. In addition, the width of both components behaves in the same way as the content. This means that they are now also bordered and centered in width.

# Changes Made

- Created new component <HeaderContainer>
- Changes on <AppShell>: <HeaderContainer> added. In addition, there was a command that gave the MainContainer a margin to the left. As this is no longer required, it has been deleted
- Example App: TopNavigation added 

# Related Issues

There is no ticket

# Screenshots (if applicable)

image

# Testing Instructions

The behavior can only be viewed in the example app
1. `npm i`
2. `npm run clean:cache `
3. `npm run pkg @cloudoperators/juno-app-example`

Component in Storybook
1. `npm i`
2. `npm run storybook -w @cloudoperators/juno-ui-components `

# Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (andypf) [inline packages/ui-components/src/components/AppShell/AppShell.component.tsx:45]: Please change `{topNavigation && topNavigation}` to just `{topNavigation}` for simplicity

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (andypf) [inline packages/ui-components/src/components/AppShell/AppShell.component.tsx:59]: Please change `{fullWidthContent === true ? true : false}` to just `{fullWidthContent === true}` for simplicity

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (andypf) [inline packages/ui-components/src/components/AppShell/AppShell.component.tsx:65]: ` {topNavigation && topNavigation}` -> ` {topNavigation}`

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (andypf) [inline packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:28]: Can you confirm that this statement is correct. It looks like if not fullWidth then add `jn-w-full` class

**Review by andypf — COMMENTED** [review]
Good Job! Please take alook at my comments

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (MartinS-git) [reply packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:28]: Yes, this is correct. `jn-w-full` (default) is active on smaller displays, on larger resolutions `jn-container` is active instead

File: .changeset/giant-news-flash.md
Comment (guoda-puidokaite) [inline .changeset/giant-news-flash.md:5]: Small grammar improvement:

```suggestion
Created HeaderContainer component to make PageHeader and TopNavigation sticky when scrolling the content. AppShell is also affected by this change.
```

File: apps/example/src/App.jsx
Comment (guoda-puidokaite) [inline apps/example/src/App.jsx:70]: **(Question)**

Wondering why we check for `string` value here if `embedded` has `PropTypes.bool `?
We'll be refactoring this anyway, so can leave it. But just curious. 🙈

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.stories.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/HeaderContainer/HeaderContainer.stories.tsx:14]: **(Blocking Improvement)**

Storybook doesn't pick up type `ReactNode` for children and reverts to `any`. Here's an [improvement](https://github.com/cloudoperators/juno/pull/601/commits/f1fa2903825034370c3449a14afaf62ae6f69686).

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:23]: **(Blocking Improvement)**

To improve consistency, readability and type safety 🙏: 

```suggestion
export const HeaderContainer: React.FC<HeaderContainerProps> = ({
  fullWidth = false,
  className = "",
  children = null,
  ...props
}) => {
```

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:38]: **(Blocking Improvement)**

Were trying to do [this](https://github.com/cloudoperators/juno/issues/583) for all TS components:

```suggestion
export interface HeaderContainerProps extends React.HTMLAttributes<HTMLDivElement> {
```

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:26]: **(Non-Blocking Improvement)**

In other components, we try extract the CSS classes into a variable so the component is easier to read.
Could we extract into `const containerStyles`, or something like that?

**Review by guoda-puidokaite — APPROVED** [review]
Really nice to have a component directly in TS! Super cool. 🚀

Just made some suggestions below. Feel free to only fix the blocking and just consider the rest. 💪 Thank You so much.

**Review by guoda-puidokaite — CHANGES_REQUESTED** [review]
My brain! 🤯 I meant to request changes above.

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (MartinS-git) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:64]: You are absolutely right with the empty className. I will remove this. I would rather not touch the other point, as I am not familiar with the background either.

File: apps/example/src/App.jsx
Comment (MartinS-git) [reply apps/example/src/App.jsx:70]: Unfortunately, I can't answer that either

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:64]: Thank You so much. That's no problem at all, it's not blocking. I'm looking at the components as a whole and ask to understand whether we could add some things to the improvements EPIC etc. But we should keep changes to requirements and lines touched in PR. 🙈

File: packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx
Comment (MartinS-git) [reply packages/ui-components/src/components/HeaderContainer/HeaderContainer.component.tsx:26]: I'm not quite sure about this point. Here there are already `const headerContainerStyles` with the main styles..

File: apps/example/src/App.jsx
Comment (andypf) [reply apps/example/src/App.jsx:70]: The Example App was previously integrated into web pages via a <script> tag using a widget loader, which only allowed string-based props to be passed. However, we have since moved away from using the widget loader and now utilize direct dynamic imports. As a result, the reliance on string-based props is a legacy from the previous approach.

File: packages/ui-components/src/components/HeaderContainer/index.ts
Comment (barsukov) [inline packages/ui-components/src/components/HeaderContainer/index.ts:6]: Do we need to export it with a type?

Comment (MartinS-git) [reply packages/ui-components/src/components/HeaderContainer/index.ts:6]: I'm not sure about that. I have looked at the other containers in the AppShell and except for `MainContainerInner`, `type` is used everywhere. In my research, I have not found any reason that speaks against it. As an expert in this topic, can you please clarify?

**Review by barsukov — CHANGES_REQUESTED** [review]
I have a question of unnecessary check on a stirng.Maybe we could discuss it.

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (MartinS-git) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:60]: I can't tell you that either. It may be a fallback solution from earlier versions. I guess Esther or @franzheidl  have found a requirement here?

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:60]: I also wondered about it here 👀 https://github.com/cloudoperators/juno/pull/607#discussion_r1843645481

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:60]: Wonder if it's a sign 😅

File: packages/ui-components/src/components/HeaderContainer/index.ts
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/HeaderContainer/index.ts:6]: I was also wondering... now that you brought it up, @barsukov... I tried without, looks like we need to use the keyword. 🤷‍♀️
Screenshot 2024-11-19 at 11 21 51

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/HeaderContainer/index.ts:6]: But I guess this is re-exporting.

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/HeaderContainer/index.ts:6]: Should we be doing this in all components we're migrating, @barsukov? Exporting the type from the component index file.

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:1]: I can do so tomorrow. But you don't have to wait for me, feel free to get a second review and dismiss my blocking request.

**Review by barsukov — APPROVED** [review]
LGTM

File: packages/ui-components/src/components/HeaderContainer/index.ts
Comment (barsukov) [reply packages/ui-components/src/components/HeaderContainer/index.ts:6]: I think in general this concept of index files are anoying in a way I will just prefer to do export types directly, but it is ok. It looks like it is a typescript setting. So leave the type

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/AppShell/AppShell.component.tsx:49]: **(Non-Blocking Improvement)**

We already check `if (!pageHeader && !topNavigation)`

```suggestion
        {typeof pageHeader === "string" ? <PageHeader heading={pageHeader} /> : pageHeader}
```

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/AppShell/AppShell.component.tsx:42]: **(Question/Non-Blocking Improvement)**

I see that this makes the type a little more reusable. 

However, we usually define types directly. One benefit is that the function is understandable to the developer without searching for the types. Makes things more readable.

Wdyt? @barsukov @MartinS-git

**Review by guoda-puidokaite — COMMENTED** [review]
Looks great, thanks for the improvements!!! And sorry for the delay due to training. 🙏
Last few improvements. Last type related one would be the most important.

Otherwise, ready to approve. 🚀

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (barsukov) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:42]: That was my suggestion to reuse a types, kinda best practices in this case :)

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (barsukov) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:61]: Not sure about it, could you double check?

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:61]: Wait, nvm, oopsy. We only check both `true`. 🤯

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:49]: Wait, nvm, oopsy. We only check if both `true`. 🤯

File: packages/ui-components/src/components/AppShell/AppShell.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/AppShell/AppShell.component.tsx:61]: Should we reuse then?

```suggestion
          renderHeaderContainer(undefined, topNavigation)
```

**Review by guoda-puidokaite — APPROVED** [review]
Looks good! Thank You! 💪

**Pull request merged**
